lib/mayaUsd/ufe/UsdContextOps.cpp

```diff
@@ -1154,33 +1042,78 @@ Ufe::ContextOps::Items UsdContextOps::getItems(const Ufe::ContextOps::ItemPath&
+            MGlobal::executeCommand("mayaUsdGetMaterialsFromRenderers", materials);
+
+            for (const auto& materials : materials) {
+                // Expects a string in the format "renderer/Material Name|Material Identifier".
+                MStringArray rendererAndMaterial;
+                MStatus      status = materials.split('/', rendererAndMaterial);
+                if (status == MS::kSuccess && rendererAndMaterial.length() == 2) {
+                    renderersAndMaterials.emplace(
+                        std::string(rendererAndMaterial[0].asChar()), rendererAndMaterial[1]);
+                }
             }
-        } else if (itemPath.size() == 2u && itemPath[1] == kAssignNewUsdMaterialItem) {
-            items.emplace_back(
-                kAssignNewUsdPreviewSurfaceMaterialItem, kAssignNewUsdPreviewSurfaceMaterialLabel);
-        } else if (itemPath.size() == 2u && itemPath[1] == kAssignNewMaterialXMaterialItem) {
-            for (auto&& menuEntry : getMaterialXSurfaceShaders()) {
-                items.emplace_back(menuEntry._identifier, menuEntry._label);
+
+            if (itemPath.size() == 1u) {
+                // Populate list of known renderers (first menu level).
+                for (auto it = renderersAndMaterials.begin(), end = renderersAndMaterials.end();
+                     it != end;
+                     it = renderersAndMaterials.upper_bound(it->first)) {
+                    items.emplace_back(it->first, it->first, Ufe::ContextItem::kHasChildren);
+                }
+            } else if (itemPath.size() == 2u) {
+                // Populate list of materials for a given renderer (second menu level).
+                const auto range = renderersAndMaterials.equal_range(itemPath[1]);
+                for (auto it = range.first; it != range.second; ++it) {
+                    MStringArray materialAndIdentifier;
+                    // Expects a string in the format "Material Name|MaterialIdentifer".
+                    MStatus status = it->second.split('|', materialAndIdentifier);
+                    if (status == MS::kSuccess && materialAndIdentifier.length() == 2) {
+                        items.emplace_back(
+                            materialAndIdentifier[1].asChar(), materialAndIdentifier[0].asChar());
+                    }
+                }
+            }
+        } else if (itemPath[0] == kAssignExistingMaterialItem) {
+            std::multimap<std::string, MString> pathsAndMaterials;
+            MStringArray                        materials;
+            MString                             script;
+            script.format(
+                "mayaUsdGetMaterialsInStage \"^1s\"",
+                Ufe::PathString::string(fItem->path()).c_str());
+            MGlobal::executeCommand(script, materials);
+
+            for (const auto& material : materials) {
+                MStringArray pathAndMaterial;
+                MStatus      status = material.split('/', pathAndMaterial);
```

Comment:
Why not a reverse find for the last "/" that will allow you to split the string in two at the right location?